qiskit_experiments/curve_analysis/curve_data.py

```diff
@@ -137,3 +139,205 @@ class ParameterRepr:
+
+        Returns:
+            Formatted value.
+
+        Raises:
+            AnalysisError: When value is invalid format.
+        """
+        if value is None:
+            return None
+
+        try:
+            return float(value)
+        except (TypeError, ValueError) as ex:
+            raise AnalysisError(f"Input value {value} is not valid initial guess. ") from ex
+
+
+class Boundaries(OptionsDict):
+    """Dictionary providing a validation for boundaries."""
+
+    @staticmethod
+    def format(value: Any) -> Any:
+        """Validate if value is a min-max value tuple.
+
+        Args:
+            value: New value to assign.
+
+        Returns:
+            Formatted value.
+
+        Raises:
+            AnalysisError: When value is invalid format.
+        """
+        if value is None:
+            return None
+
+        try:
+            minv, maxv = value
+            if minv >= maxv:
+                raise AnalysisError(
+                    f"The first value is greater than the second value {minv} >= {maxv}."
+                )
+            return float(minv), float(maxv)
+        except (TypeError, ValueError) as ex:
+            raise AnalysisError(f"Input boundary {value} is not a min-max value tuple.") from ex
+
+
+# pylint: disable=invalid-name
+class FitOptions:
+    """Collection of fitting options.
+
+    This class is initialized with list of parameter names used in the fit model
+    and associated default values provided by users.
+
+    This class is hashable, and generates fitter keyword arguments.
+    """
+
+    def __init__(
+        self,
+        parameters: List[str],
+        default_p0: Optional[Union[Iterable[float], Dict[str, float]]] = None,
+        default_bounds: Optional[Union[Iterable[Tuple], Dict[str, Tuple]]] = None,
+        **extra,
+    ):
+        self._p0 = InitialGuesses(parameters, default_p0)
+        self._bounds = Boundaries(parameters, default_bounds)
+        self._extra = extra
+
+    def __hash__(self):
+        return hash((self._p0, self._bounds, tuple(sorted(self._extra.items()))))
+
+    def __eq__(self, other):
+        if isinstance(other, FitOptions):
+            checks = [
+                self._p0 == other._p0,
+                self._bounds == other._bounds,
+                self._extra == other._extra,
+            ]
+            return all(checks)
+        return False
+
+    def add_extra_options(self, **kwargs):
+        """Add more fitter options."""
+        self._extra.update(kwargs)
+
+    def copy(self):
+        """Create copy of this option."""
+        return FitOptions(list(self._p0.keys()), dict(self._p0), dict(self._bounds), **self._extra)
+
+    @property
+    def p0(self) -> InitialGuesses:
+        """Return initial guess dictionary."""
+        return self._p0
+
+    @p0.setter
+    def p0(self, new_dict: Dict[str, float]):
+        """Set new dictionary. Only non-assigned value is updated."""
+        self._p0.set_if_empty(**new_dict)
+
```

Comment:
Do we really need these setters? They sort of mask the fact that `set_if_empty` is called. Looking at the rest of the code, e.g. DragAnalysis I see a lot of `opt.p0.set_if_empty` calls which makes me believe that we do not need the `p0` setter here and the bounds setter below.
```suggestion
```

Reply:
Okey, this is to prevent 

```python
opt.p0 = {"a": 0, "b": 1}
```

since `OptionsDict` looks like a dictionary (kind of intended) and some community developer who is not familiar with it may set raw dict and disable all validations (no validation is performed without `OptionsDict` class now). I'll just remove `set_if_empty` then.

Reply:
Done in https://github.com/Qiskit/qiskit-experiments/pull/390/commits/4628d2a7002637cfb8a215fc960e0197a719f797

Reply:
Or alternative approach would be turning p0 and bounds into private member (`__`) and removing the setter. This causes error when something is directly set to the guess and bounds.

Reply:
Approach changed in https://github.com/Qiskit/qiskit-experiments/pull/390/commits/0681a919a82b4c38840cc673b5224e9c15d35366 The setters are finally removed. A code something like `opt.p0 = {"a": 0}` confuses the linter and we need to write `pylint: disable=no-member` everywhere by accepting dict type (linter assumes p0 is dict rather than OptionsDict, and causes no-member error when it finds set_if_empty method in other places). This is bit awkward and we can write code simplify without removing setters. Now the code directly sets a dict to options is not allowed.